src/coordinated_workers/coordinator.py

```diff
@@ -865,3 +911,352 @@ def _setup_charm_tracing(self):
+                NginxLocationConfig(
+                    path=PROXY_WORKER_TELEMETRY_PATHS["remote_write_endpoints"].format(
+                        unit=unit_name_sanitized
+                    ),
+                    backend=upstream_name,
+                    backend_url=parsed_address.path,
+                    upstream_tls=parsed_address.scheme.endswith("s"),
+                    modifier="=",  # exact match, match only the remote write path
+                    is_grpc=False,
+                )
+            )
+
+        return upstreams, locations
+
+    def _generate_tracing_urls_nginx_config(
+        self,
+    ) -> Tuple[List[NginxUpstream], List[NginxLocationConfig]]:
+        """Generate nginx upstreams and locations for routing charm and workload tracing."""
+        upstreams: List[NginxUpstream] = []
+        locations: List[NginxLocationConfig] = []
+        created_upstreams: Set[str] = set()
+
+        tracing_types = [
+            (
+                "charm",
+                self._charm_tracing_receivers_urls,
+                PROXY_WORKER_TELEMETRY_PATHS["charm_tracing_receivers_urls"],
+            ),
+            (
+                "workload",
+                self._workload_tracing_receivers_urls,
+                PROXY_WORKER_TELEMETRY_PATHS["workload_tracing_receivers_urls"],
+            ),
+        ]
+
+        for tracing_type, receivers_urls, path_template in tracing_types:
+            for protocol, address in receivers_urls.items():
+                parsed_address = urlparse(address)
+                upstream_name = (
+                    f"{PROXY_WORKER_TELEMETRY_UPSTREAM_PREFIX}-{tracing_type}-{protocol}"
+                )
+
+                # Create upstream if we haven't already
+                if upstream_name not in created_upstreams:
+                    upstreams.append(
+                        NginxUpstream(
+                            name=upstream_name,
+                            port=parsed_address.port,  # type: ignore
+                            address_lookup_key=upstream_name,
+                        )
+                    )
+                    created_upstreams.add(upstream_name)
+
+                # Create tracing location
+                locations.append(
+                    NginxLocationConfig(
+                        path=path_template.format(protocol=protocol),
+                        backend=upstream_name,
+                        backend_url=parsed_address.path,
+                        rewrite=[
+                            f"^{path_template.format(protocol=protocol)}(.*)",
+                            "/$1",
+                            "break",
+                        ],  # strip the custom prefix before redirecting
+                        upstream_tls=parsed_address.scheme.endswith("s"),
+                        is_grpc=False,  # FIXME: GRPC must be allowed. Create an issue and link here.
+                    )
+                )
+
+        return upstreams, locations
```

Comment:
we're adding >300 lines of code to a module that's already nearly 1000 LOC long. I think we should put all of the 'proxy stuff' in a dedicated module.